### PR TITLE
LPS-110163 Improve usability of <react:component> taglib

### PR DESCRIPTION
**Pull request by wincent**
When we originally made the `render` wrapper in `frontend-js-react-web` in 31f9beca59c7ac9b1c42d3e34a842167cb8313d7 we tried to keep the API relatively close to [the `ReactDOM.render` API](https://reactjs.org/docs/react-dom.html#render) that we're wrapping:

    render(reactElement, container, callback)

Later on (e77295d3d2e6ebf0010fdaccf21d09a5d5d6a2d4) we changed the signature to:

    render(renderFunction, renderData, container)

ie. instead of a React Element (eg. `<Foo />`), you would pass a function that returns an element.

All of this was motivated by a desire to make it easy to mount React components from JSP files using the `<react:component>` taglib, as [explained here](https://github.com/brianchandotcom/liferay-portal/pull/77801).

However, in practice, people are frequently confused by this pattern, running into errors like "Invalid hook call" if they pass a function that uses hooks:

    render(FunctionThatUsesHooks, renderData, container)

They can avoid this by adding a layer of indirection, but it is not intuitive to do so:

    render(() => <FunctionThatUsesHooks />, renderData, container)

And note that if the want to actually make use of the `renderData` in an example like that, they have to propagate it through manually in a repetitive way (if they are calling `render` directly instead of using the `<react:component>` taglib):

    render(data => {
        return (
          <FunctionThatUsesHooks
            someProp={data.foo}
          />
        );
      },
      renderData,
      container
    );

or (if they are being called from the taglib):

    export default function(props) {
      return <FunctionThatUsesHooks someProp={props.foo} />;
    }

I think we can get rid of this confusion and reduce the indirection by making `render` a little more liberal in what it accepts. Now in addition to accepting:

- A "render function" (a function that returns an element).

It also accepts:

- A component function (ie. no layer of indirection); or:
- An element (again, no layer of indirection).

In other words, given a functional component, `FooComponent`, all of these are now valid:

    render(FooComponent, data, container);
    render(() => <FooComponent />, {}, container);
    render(<FooComponent />, data, container);

To show the improvement, in this commit I remove indirection from all the call sites where it wasn't needed. There are still places where it is useful, and I left those untouched; in all of those they do some kind of prop-massaging before calling the actual component -- ie. things like this:

    export default function(({props, context}) {
      return (
        <Context.Provider value={context}>
          <App {...props} />
        </Context.Provider>
    }

### CODE REVIEW

**Comment by wincent** [comment]
@jbalsas I have only lightly tested this at this point, but another set of eyes wouldn't hurt. Tomorrow I'll actually do a bit of QA on it. Do you think this makes sense?

**Comment by wincent** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test:stable

File: modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.es.js [hunk 1]
```diff
@@ -58,17 +62,19 @@ export default function render(renderFunction, renderData, container) {
 			}
 		);
 
+		const Component = typeof renderable === 'function' ? renderable : null;
+
 		// eslint-disable-next-line liferay-portal/no-react-dom-render
 		ReactDOM.render(
 			<ClayIconSpriteContext.Provider value={spritemap}>
-				{renderFunction(renderData)}
+				{Component ? <Component {...renderData} /> : renderable}
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.es.js]: So, if I'm reading this correctly... when `renderable` is already a Component we just use it. In the case it's a function, we assign `renderable` to `Component` and then render it passing `renderData`?

I think the `Component === null when renderable is Component` confuses me, but it's probably just me :)

**Comment by jbalsas** [comment]
LGTM @wincent! Do you feel this is safe to push forward or would you rather wait for GA2?

Seeing `stable` green I'd be inclined to just send it over... wdt?

File: modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.es.js (hunk 1 above)
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.es.js]: Yeah, it's a bit tricky. `renderable` could be:

-   A functional component:

    ```js
    const MyFancyGreeting = props => <span>hi</span>`;
    ```

- A function that returns an element:

    ```js
    function FancyGreeting(renderData) {
        SomeImperativeSetUp();
        return (
            <Context value={renderData.context}>
                <Greeting {...renderData.props} />
            </Context>
        );
    }
    ```

-   A React element:

    ```js
    <MyGreeting />
    ```

Note that those first two look awfully alike, and from the outside, they look identical (all we really know about them is that they have `typeof === 'function'`). But even though the second one technically looks just like a functional component and even satisfies the definition of one, conceptually, it is subtly different: "do arbitrary set up code then return a React Element".

So, when we do `<Component {...renderData} />` we may be effectively "double-wrapping" the component, and *that's* what avoids the "invalid hook call" errors. The total effect of the two ternaries is to ensure that we always render a React element (ie. the thing you get from `React.createElement`).

**Comment by wincent** [comment]
Closing this and will forward after testing it. Thanks for the 👀

File: modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.es.js (hunk 1 above)
Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.es.js]: Makes sense, thanks for the explanation!! 🏹 (bow)

**Comment by wincent** [comment]
Forwarded: https://github.com/brianchandotcom/liferay-portal/pull/86171
